numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -3653,14 +3653,21 @@ NPY_NO_EXPORT void
 
     if (is1b && is2b)
     {
-        double sum = 0.;  /* double for stability */
+        double sum = -0.;  /* double for stability */
+        double result;
 
         while (n > 0) {
             CBLAS_INT chunk = n < NPY_CBLAS_CHUNK ? n : NPY_CBLAS_CHUNK;
 
-            sum += CBLAS_FUNC(cblas_@prefix@dot)(chunk,
+            result = CBLAS_FUNC(cblas_@prefix@dot)(chunk,
                                      (@type@ *) ip1, is1b,
                                      (@type@ *) ip2, is2b);
+            if (npy_signbit(((@type@ *)ip1)[0]) &&
+                ((@type@ *)ip1)[0] == 0 &&
+                !npy_signbit(((@type@ *)ip2)[0])) {
+                result = -0.0;
+            }
```

Comment:
This (and the earlier thing the same way probably), doesn't look right.  `np.dot([-0., 0.], [0., 0.])` should return a positive 0 and I don't think it will here.

I was happy with the idea (and agree that it seems more correct) when I though this is just about initialization inputs with `-0.0` (unless empty).
And initialization is about where I would want to draw the line of contortions to do to get this right.

But...  I can't find a `res = alpha + x.T @ y` in level 1 BLAS.  So, this isn't possible beyond (potentially, very unlikely) inspecting all input values and redoing most of the calculation.  (Only when the result is `0.0`).  To me, that seems unacceptable considering the magnitude of the issue ;).

So, I am tempted to close the issue with:  Pick it up with BLAS implementors, the best I can offer is a "meh, NumPy dev's somewhat agree that -0.0 should be returned in this cases considering IEEE rules".
(Next thing, do CPU dot intrinsics actually get it right?  My guess is yes, but that is probably where BLAS implementors would draw the line.)

Reply:
Sorry, the whole comment refers only the 1-D dot product, I guess.  2-D dot product *can* initialize the output correctly, and I am happy to pursue that a bit more.

(Maybe there is a way to abuse level 2-blas without any slow-down to get this, maybe that would be a point...)